### ISSUE TITLE
Increase flexiblility of MPI task partition patterns

### ISSUE THREAD

**Issue by yizhang-cae** [stan-dev/math#948]
#### Summary:
Allow more partition patterns in MPI jobs, instead of single equally-divided job allocation pattern.

#### Description:
`map_rect` only allows equally-divided pattern for MPI job allocation. That is, in `mpi_map_chunks` the number of jobs in each node is calculated as `total_number_of_jobs/total_number_of_nodes`. This is assuming each job is small enough to be efficiently handled by a single node. For many problems this assumption is not met, such examples include user-supplied PDE solution routine. For such an application, we may only have a single job for each chain, but that job requires `N` nodes to process, where `N>>1`. Current `mpi_map_chunks` function will map this single job to a single node, and leave the rest `N-1` node idle, instead of delegating the whole `N` nodes to the task(how the `N` nodes are used to perform the task depends on application).

A related issue is #940 , where we allow specifying communicator to construct `mpi_cluster`, so that a group of dedicated nodes can be used in a task-specific communication group.

For example, we may look at an input like the following.
```Stan
functions {
   real[] user_function_based_on_metis(...);
  }
}
parameters {
real theta;
...
}
....
model {
  ....
  target += sum(map_rect(user_function_based_on_metis, beta, theta, xs, ys));
}
```
Since there is only one `theta`, no matte how many nodes we throw in, only the root will be used to perform the task of evaluating `user_function_based_on_metis`, the heavy-lifting function uses `theta` to solve a problem using MPI partition produced by METIS. What SHOULD happen is ALL the nodes issued are used to evaluate this function.

TL;DR
Current `mpi_map_chunks` pattern is too rigid, we need fix it.
#### Current Version:
v2.18.0

**Comment by wds15** [comment]
Yes, we need this. I agree. I don‘t have a good idea at the moment. I aimed at making a first simple as possible solution, but we should improve. What was important in the design of the logic is that we always assign with the same workload things to the same nodes. However, we should make that more flexible and maybe even think about some more clever scheduling things. So if the workload distributes very uneven over the cluster, then we should probably change the job to node assignment dynamically for the next iteration.

**Comment by yizhang-cae** [comment]
@wds15 mentioned in the weekly meeting that this is for PDE applications. Though PDE will definitely benefit from it, it goes beyond that. Whenever a user passes an expensive function to `map_rect`, he may want to control how many MPI resources are allocated to that function. That function could be plain Stan, not just an external PDE solution process.

**Comment by seantalts** [comment]
I'm trying to wrap my head around this, let me try to paraphrase and see if I have it right:
You have something like a `user_function_based_on_metis` that will use many MPI nodes under the hood and you want to allow it to be used from within `map_rect`?

How is MPI within `user_function_based_on_metis` configured? What does `map_rect` have to do with it if you're just passing a single row to it? I guess I'm not seeing why you'd use `map_rect` at all for this if you just have one item you want to map over - wouldn't you rather set up MPI for `user_function_based_on_metis` some other way?

**Comment by yizhang-cae** [comment]
I can, because the user routine can manipulate MPI as much as it wants. But that beats the purpose of `map_rect` of facilitating the map-reduce pattern. Say the user wants to making inference on `theta` by doing
```Stan
sum(map_rect(user_function, theta, user_param, xs, ...))
```
and `user_function` is expensive. He'll have to break up `xs`, loop it over, manage MPI resources, run `user_function`, and collect the results. Basically he ends up reproducing most of the task of `map_rect`, and this will happen to next user with similar situation.

**Comment by seantalts** [comment]
Sorry, still not getting it... In your example above I thought `xs` was just one item. So if `xs` has many things and `user_function` is expensive and you want to map over it, that seems like it's perfect for `map_rect`. Is the issue that `user_function` calls some other `user_function` that it would like to also map over? i.e. that you're looking for nested `map_rect` calls?

**Comment by yizhang-cae** [comment]
There several ways. One is through #939 I proposed. We can have `user_function` only run by nodes within a given communicator.

**Comment by seantalts** [comment]
The lecture note linked in that issue in particular was really interesting! I feel like I have a much better understanding now, thank you. 

So I think I see how you might want to add arguments to `mpi_cluster` in #939 to allow you to give a subset of your MPI cluster to Stan, broadly. But how do you want that to interact with `map_rect`? One option that I see:
1. Of the global Stan Math mpi_cluster comm, have `map_rect` internally assign `max(comm.size(), length(xs)` nodes to its processes and, somehow, pass a comm containing the rest of the unused nodes to `user_function`. Would you create a new `map_rect` that accepts a `user_function` mapping user defined function with an additional `comm` argument?

**Comment by bob-carpenter** [comment]
These were the links in the issue to which Sean was pointing:

http://www.mcs.anl.gov/research/projects/mpi/mpi-standard/mpi-report-1.1/node114.htm

A lecture note
https://www.archer.ac.uk/training/course-material/2015/10/AdvMPI_EPCC/S1-L04-Split-Comms.pdf
